src/services/Functional/User.js

```diff
@@ -1,51 +1,41 @@
-      }
-    }
+  async reAuthenticate(store) {
+    // set re Authentication state as in progress
+    store.dispatch("auth/setReAuthenticationState", "in-process");
 
-    // if the reAuthentication call is made while the user has been logged out
-    // then reject the promise
-    if (!store.getters["auth/isAuthenticated"])
-      return Promise.reject("User is not set!");
+    // create a promise and store it's resolver in a separate variable
+    let reAuthenticationPromiseResolver;
+    let reAuthenticationPromise = new Promise((resolve) => {
+      reAuthenticationPromiseResolver = resolve;
+    });
 
-    // set re-authentication as being in process
-    store.dispatch("auth/setReAuthenticationState", true);
+    // save the above created promise and it's resolver into the store
+    store.dispatch("auth/setReAuthenticationPromise", reAuthenticationPromise);
+    store.dispatch(
+      "auth/setReAuthenticationPromiseResolver",
+      reAuthenticationPromiseResolver
+    );
 
-    // start the refreshAccessToken call and save that promise
-    // in a variable called "reAuthenticationPromise"
-    const reAuthenticationPromise = UserAPIService.refreshAccessToken()
-      .then((response) => {
-        // once the refreshAccessToken call resolves, set the new access token in store
-        // set isReAuthenticating to false
-        // and unset the reAuthenticationPromise that was saved in store
-        // and resolve the promise so other waiting calls can proceed
+    try {
+      // try refreshing the access token
+      let response = await UserAPIService.refreshAccessToken();
+      if (response != undefined && response.data != undefined) {
+        // set the recieved access token into the store
+        // and mark the re authentication process as completed.
+        // We will also resolve the pending reAuthenticationPromise so the code
+        // that is waiting for the promises to resolve can proceed
         store.dispatch("auth/setAccessToken", response.data);
-        store.dispatch("auth/setReAuthenticationState", false);
-        store.dispatch("auth/unsetReAuthenticationPromise");
-        return Promise.resolve(true);
-      })
-      .catch((error) => {
-        return Promise.reject(error);
-      });
-
-    // Until the promise stored in "reAuthenticationPromise" is resolved,
-    // store this promise into vuex and return from the function.
-    store.dispatch("auth/setReAuthenticationPromise", reAuthenticationPromise);
-    return reAuthenticationPromise;
+        store.dispatch("auth/setReAuthenticationState", "completed");
+        store.state.auth.reAuthenticationPromiseResolver();
+        return response.data;
+      }
+    } catch (e) {
+      // an error occured in the process of refreshing the access token.
+      // We'll mark the re authentication process as dropped and throw an error
+      store.dispatch("auth/setReAuthenticationState", "dropped");
+      throw new Error(e);
```

Comment:
we can use a PromiseRejector as well and use Promise.reject as well right? Would keep it consistent with the way we've been handling errors in general, no?

Reply:
Again, as this is an async function, it will always return a promise (refer to the image attached above in one comment).

If it was not an async function, we would've written Promise.reject(...).
It's not about being consistent actually here. This is the correct way.

Reply:
Didn't understand what you mean to say here. Will call!